### PR TITLE
[FLINK-13222] [runtime] Add documentation for failover strategy option

### PR DESCRIPTION
**Pull request by zhuzhurk**
## What is the purpose of the change

*The PR adds documentation for failover strategy options, including "full" and "region".*

## Brief change log

  - *Add documentation for each failover strategy option, elaborating the failover strategy.*

## Verifying this change

This change is a trivial documentation refinement without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java [hunk 1]
```diff
@@ -106,17 +106,21 @@
...
-				.text("This option specifies how the job computation recovers from task failures. " +
+				.text("This option specifies the failover strategy, i.e. " +
+					"how the job computation recovers from task failures. " +
 					"Accepted values are:")
 				.list(
-					text("'full': Restarts all tasks."),
-					text("'individual': Restarts only the failed task. Should only be used if all tasks are independent components."),
-					text("'region': Restarts all tasks that could be affected by the task failure.")
+					text("'full': Use RestartAllStrategy to restarts all tasks to recover the job."),
+					text("'region': Use AdaptedRestartPipelinedRegionStrategyNG to restart all tasks " +
```
Comment (GJL) [inline flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java]: I would avoid using class names, such as `AdaptedRestartPipelinedRegionStrategyNG`, here. We should describe their behavior only. 

Another idea that @zentol had, is to extend the section about restart strategies [1] with more detailed information about the failover strategies (`full` and `region`). 

[1] https://ci.apache.org/projects/flink/flink-docs-master/dev/restart_strategies.html

Comment (zhuzhurk) [reply flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java]: Ok. Let's remove the class name and just describe the behavior here.

I think adding more information for different failover strategies is a very good idea. In this PR I'm trying to make the documentation easy to understand, but seems it's a bit hard to make it very clear if we do not want to add too many words for the option documentation.
Just not sure if the RestartStrategy page is the proper place? As the failover strategy and restart strategy are somehow independent.

Comment (zentol) [reply flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java]: From a users perspective restart and failover strategies are connected quite closely, with one defining the "how" and the other defining "how often".
From a technical perspective they also aren't independent, since if restarts are disabled the failover strategy will not kick in.

Comment (zhuzhurk) [reply flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java]: Just feel that adding a "Failover Strategies" section in "Restart Strategies" might be a bit confusing to users, indicating that failover strategy is part of restart strategy.  

Would it be better to re-org the page as below?
**Task Failure Recovery**
  +- Restart Strategies
     +- Fixed Delay Restart Strategy
     +- Failure Rate Restart Strategy
     +- ...
  +- Failover Strategies
     +- Restart All Strategy
     +- Restart Pipelined Region Strategy

Comment (GJL) [reply flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java]: I would be ok with changing the heading of the page. If you are changing the file name, I'd introduce a redirect so that we do not break existing hyperlinks (see https://github.com/apache/flink/tree/9cdae7307e9be42be473a22d5a3e924c159ca12d/docs/redirects)

Comment (zhuzhurk) [reply flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java]: Thanks for the reminding. Will take it this way.

File: docs/dev/task_failure_recovery.md [hunk 2]
```diff
@@ -264,4 +268,44 @@ The cluster defined restart strategy is used.
...
+        <td>full</td>
+    </tr>
+    <tr>
+        <td>Restart pipelined region</td>
+        <td>region</td>
+    </tr>
+  </tbody>
+</table>
+
+### Restart All Strategy
```
Comment (GJL) [inline docs/dev/task_failure_recovery.md]: I have the feeling that a user could confuse this with a *Restart Strategy*.
How about *Restart All Failover Strategy*? If you have other ideas, let me know.

File: docs/dev/task_failure_recovery.zh.md [hunk 3]
```diff
@@ -260,4 +265,42 @@ env.setRestartStrategy(RestartStrategies.noRestart())
 这对于启用了 checkpoint 的流处理程序很有帮助。
 如果没有定义其他重启策略，默认选择固定延时重启策略。
 
+## 故障恢复策略
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply docs/dev/task_failure_recovery.zh.md]: Sure.

File: docs/dev/task_failure_recovery.md [hunk 4]
```diff
@@ -264,4 +268,44 @@ The cluster defined restart strategy is used.
...
+
+With this strategy, all tasks in the job will be restarted to recover from a task failure.
+
+### Restart Pipelined Region Strategy
+
+With this strategy, tasks to restart depend on the regions to restart.
+A region is defined by this strategy as tasks that communicate via pipelined data exchange.
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply docs/dev/task_failure_recovery.md]: Good idea.

File: docs/dev/task_failure_recovery.md (hunk 2 above)
Comment (zhuzhurk) [reply docs/dev/task_failure_recovery.md]: Yes. The naming for failover strategies are confusing with restart strategies.
I don't have any good ideas yet. Let's make it _Restart All Failover Strategy_ here to reduce possible confusion.

File: docs/dev/task_failure_recovery.md [hunk 5]
```diff
@@ -1,5 +1,5 @@
 ---
-title: "Restart Strategies"
+title: "Task Failure Recovery"
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply docs/dev/task_failure_recovery.md]: Agree. The batch fault tolerance page is suggesting a deprecated way to configure restart strategies. Deleting it can also help with further deprecating it.

File: flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java [hunk 6]
```diff
@@ -106,17 +106,20 @@
...
-	@Documentation.ExcludeFromDocumentation("The failover strategy feature is highly experimental.")
 	public static final ConfigOption<String> EXECUTION_FAILOVER_STRATEGY =
 		key("jobmanager.execution.failover-strategy")
 			.defaultValue("full")
 			.withDescription(Description.builder()
 				.text("This option specifies how the job computation recovers from task failures. " +
 					"Accepted values are:")
 				.list(
-					text("'full': Restarts all tasks."),
-					text("'individual': Restarts only the failed task. Should only be used if all tasks are independent components."),
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java]: Agree.

File: docs/dev/task_failure_recovery.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply docs/dev/task_failure_recovery.md]: The ExecutionMod is for DataSet job only.
Considering the streaming job and the batch table job  introduced in 1.9, will elaborate more about the data exchanges for different job types.

File: docs/dev/task_failure_recovery.zh.md [hunk 7]
```diff
@@ -260,4 +265,49 @@ env.setRestartStrategy(RestartStrategies.noRestart())
...
+在全图重启故障恢复策略下，Task 发生故障时会重启作业中的所有 Task 进行故障恢复。
+
+### 基于 Region 的局部重启故障恢复策略
+
+本策略会以 Region 为粒度来决定需要重启的 Task。
+
+此处 Region 指以 Pipelined 形式进行数据交换的 Task 集合。
+- DataStream 和 流式 Table 作业的所有数据交换都是 Pipelined 形式的。
+- 批处理式 Table 作业的所有数据交换都是 Batch 形式的。
+- DataSet 作业中的数据交换形式会根据 [ExecutionConfig]({{ site.baseurl }}/dev/execution_configuration.html)
```
Comment (wuchong) [inline docs/dev/task_failure_recovery.zh.md]: Use the Chinese version link which is start with "zh": `({{ site.baseurl }}/zh/dev/execution_configuration.html) `

**Comment by wuchong** [comment]
Thanks for updating the Chinese documentation as well. It looks good to me.  I only left a minor comment about the link.

File: docs/dev/task_failure_recovery.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply docs/dev/task_failure_recovery.md]: I opened a ticket for it: https://issues.apache.org/jira/browse/FLINK-13336 .
Can start working on it once this PR is merged.

File: docs/dev/task_failure_recovery.md [hunk 8]
```diff
@@ -264,4 +268,51 @@ The cluster defined restart strategy is used.
...
+- Types of data exchanges in a DataSet job is decided with the 
+  [ExecutionMode]({{ site.javadocs_baseurl }}/api/java/org/apache/flink/api/common/ExecutionMode.html) set in 
+  [ExecutionConfig]({{ site.baseurl }}/dev/execution_configuration.html).
+
+Regions to restart are decided as below:
+1. The region containing the failed task should be restarted.
+2. If a result partition is not available while it is required by a region that will be restarted,
+   the region producing the result partition should be restarted as well.
+3. If a region is to be restarted, all of its consumer regions should also be restarted. This is to guarantee
+   data consistency. Because nondeterministic processing or partitioning can result in that a partition
```
Comment (GJL) [inline docs/dev/task_failure_recovery.md]: *[...] data consistency because nondeterministic [...]*

File: docs/dev/task_failure_recovery.md [hunk 9]
```diff
@@ -264,4 +268,51 @@ The cluster defined restart strategy is used.
...
+
+A region is defined by this strategy as tasks that communicate via pipelined data exchanges.
+- All data exchanges in a DataStream job or Streaming Table job are pipelined.
+- All data exchanges in a Batch Table job are batched.
+- Types of data exchanges in a DataSet job is decided with the 
+  [ExecutionMode]({{ site.javadocs_baseurl }}/api/java/org/apache/flink/api/common/ExecutionMode.html) set in 
+  [ExecutionConfig]({{ site.baseurl }}/dev/execution_configuration.html).
+
+Regions to restart are decided as below:
+1. The region containing the failed task should be restarted.
```
Comment (GJL) [inline docs/dev/task_failure_recovery.md]: I'd change: *should be restarted* ->  *is restarted*

File: flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java [hunk 10]
```diff
@@ -105,18 +105,27 @@
...
-					text("'region': Restarts all tasks that could be affected by the task failure.")
+					text("'full': Restarts all tasks to recover the job."),
+					text("'region': Restarts all tasks that could be affected by the task failure. " +
```
Comment (GJL) [inline flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java]: Maybe we can omit some details here, and place a hyperlink to the *Task Failure Recovery* page using `LinkElement`. WDYT?

File: docs/dev/task_failure_recovery.md [hunk 11]
```diff
@@ -264,4 +268,51 @@ The cluster defined restart strategy is used.
...
+With this strategy, tasks to restart depend on the regions to restart.
+
+A region is defined by this strategy as tasks that communicate via pipelined data exchanges.
+- All data exchanges in a DataStream job or Streaming Table job are pipelined.
+- All data exchanges in a Batch Table job are batched.
+- Types of data exchanges in a DataSet job is decided with the 
+  [ExecutionMode]({{ site.javadocs_baseurl }}/api/java/org/apache/flink/api/common/ExecutionMode.html) set in 
+  [ExecutionConfig]({{ site.baseurl }}/dev/execution_configuration.html).
+
+Regions to restart are decided as below:
```
Comment (GJL) [inline docs/dev/task_failure_recovery.md]: ***The** regions to restart [...]*

File: docs/dev/task_failure_recovery.md [hunk 12]
```diff
@@ -264,4 +268,51 @@ The cluster defined restart strategy is used.
...
+  </tbody>
+</table>
+
+### Restart All Failover Strategy
+
+With this strategy, all tasks in the job will be restarted to recover from a task failure.
+
+### Restart Pipelined Region Failover Strategy
+
+With this strategy, tasks to restart depend on the regions to restart.
```
Comment (GJL) [inline docs/dev/task_failure_recovery.md]: I'd rephrase this and add a few more details:
> This strategy groups tasks into disjoint regions. When a task failure is detected, this strategy computes the smallest set of regions that must be restarted to recover from the failure. For some jobs this can result in fewer tasks that will be restarted compared to the *Restart All Failover Strategy*.\

WDYT?

File: docs/dev/task_failure_recovery.md [hunk 13]
```diff
@@ -22,14 +22,20 @@ specific language governing permissions and limitations
...
+Restart strategies decide whether and when the failed/affected tasks can be restarted.
+Failover strategies decide which tasks should be restarted to recover the job.
 
 * This will be replaced by the TOC
 {:toc}
 
-## Overview
+## Restart Strategies
+
+Flink supports different restart strategies.
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply docs/dev/task_failure_recovery.md]: Ok.

File: flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java [hunk 14]
```diff
@@ -105,18 +105,27 @@
 
 	/**
 	 * This option specifies the failover strategy, i.e. how the job computation recovers from task failures.
+	 * The option "individual" and "region-legacy" are intentionally not included
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java]: Ah yes.

File: flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java [hunk 15]
```diff
@@ -105,18 +105,27 @@
...
+	 * as they have some known limitations or issues.
+	 * "individual" strategy only works when all tasks are not connected, in which case the "region"
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java]: Yes. This looks better.

File: docs/dev/task_failure_recovery.md [hunk 16]
```diff
@@ -264,4 +268,51 @@ The cluster defined restart strategy is used.
...
+    <tr>
+        <td>Restart pipelined region</td>
+        <td>region</td>
+    </tr>
+  </tbody>
+</table>
+
+### Restart All Failover Strategy
+
+With this strategy, all tasks in the job will be restarted to recover from a task failure.
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply docs/dev/task_failure_recovery.md]: Ok.

File: docs/dev/task_failure_recovery.md [hunk 17]
```diff
@@ -264,4 +268,51 @@ The cluster defined restart strategy is used.
...
+
+### Restart Pipelined Region Failover Strategy
+
+With this strategy, tasks to restart depend on the regions to restart.
+
+A region is defined by this strategy as tasks that communicate via pipelined data exchanges.
+- All data exchanges in a DataStream job or Streaming Table job are pipelined.
+- All data exchanges in a Batch Table job are batched.
+- Types of data exchanges in a DataSet job is decided with the
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply docs/dev/task_failure_recovery.md]: Ok.

File: docs/dev/task_failure_recovery.md (hunk 8 above)
Comment (zhuzhurk) [reply docs/dev/task_failure_recovery.md]: Ok.

File: docs/dev/task_failure_recovery.md [hunk 18]
```diff
@@ -264,4 +268,51 @@ The cluster defined restart strategy is used.
...
+
+### Restart All Failover Strategy
+
+With this strategy, all tasks in the job will be restarted to recover from a task failure.
+
+### Restart Pipelined Region Failover Strategy
+
+With this strategy, tasks to restart depend on the regions to restart.
+
+A region is defined by this strategy as tasks that communicate via pipelined data exchanges.
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply docs/dev/task_failure_recovery.md]: Ok. It's good to make it as easy to understand as possible.

The "defined by this strategy" is used to mean that in the future we may have other region strategies which may define region differently.
But in the near future looks that the _region_ is bind to the concept to _pipelined region_. 
So agree that we can simplify this definition.

File: flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java (hunk 10 above)
Comment (zhuzhurk) [reply flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java]: Ok. Let's make it this way.

File: docs/dev/task_failure_recovery.md (hunk 11 above)
Comment (zhuzhurk) [reply docs/dev/task_failure_recovery.md]: Ok.

File: docs/dev/task_failure_recovery.md (hunk 12 above)
Comment (zhuzhurk) [reply docs/dev/task_failure_recovery.md]: Ok. I think it's better.

File: docs/dev/task_failure_recovery.md (hunk 9 above)
Comment (zhuzhurk) [reply docs/dev/task_failure_recovery.md]: Here we are describing what regions should be restarted.
"is restarted" have me feel that this is happening.

Not pretty sure whether my feeling is correct. 
Please correct me if I had a wrong sense of it.

File: docs/dev/task_failure_recovery.md [hunk 19]
```diff
@@ -264,4 +268,51 @@ The cluster defined restart strategy is used.
...
+- All data exchanges in a DataStream job or Streaming Table job are pipelined.
+- All data exchanges in a Batch Table job are batched.
```
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply docs/dev/task_failure_recovery.md]: The Batch Table API is independent from the DataSet API, though they can integrate in some cases.

File: docs/dev/task_failure_recovery.md [hunk 20]
```diff
@@ -22,14 +22,19 @@ specific language governing permissions and limitations
...
+Failover strategies decide which tasks should be restarted to recover the job.
 
 * This will be replaced by the TOC
 {:toc}
 
-## Overview
+## Restart Strategies
+
+The cluster can be started with a default restart strategy which is always used when no job specific restart strategy has been defined.
+In case that the job is submitted with a restart strategy, this strategy overrides the cluster's default setting.
```
Comment (gaoyunhaii) [inline docs/dev/task_failure_recovery.md]: I think `In case that` -> `If` might be better ? IMO `In case that` means in order to let something not happen.

File: docs/dev/task_failure_recovery.md [hunk 21]
```diff
@@ -264,4 +267,54 @@ The cluster defined restart strategy is used.
...
+</table>
+
+### Restart All Failover Strategy
+
+This strategy restarts all tasks in the job to recover from a task failure.
+
+### Restart Pipelined Region Failover Strategy
+
+This strategy groups tasks into disjoint regions. When a task failure is detected, 
+this strategy computes the smallest set of regions that must be restarted to recover from the failure.
```
Comment (gaoyunhaii) [inline docs/dev/task_failure_recovery.md]: `smallest set of` -> `smallest sets of` ?

File: docs/dev/task_failure_recovery.md (hunk 20 above)
Comment (zhuzhurk) [reply docs/dev/task_failure_recovery.md]: Agreed.

File: docs/dev/task_failure_recovery.md (hunk 21 above)
Comment (zhuzhurk) [reply docs/dev/task_failure_recovery.md]: I think here we mean `one set` of regions to restart for a task failure.

Comment (GJL) [reply docs/dev/task_failure_recovery.md]: @zhuzhurk is right. A change is not required.

File: docs/dev/task_failure_recovery.md (hunk 20 above)
Comment (GJL) [reply docs/dev/task_failure_recovery.md]: It was like that before. Can be fixed in a hotfix commit.

File: docs/dev/task_failure_recovery.zh.md [hunk 22]
```diff
@@ -260,4 +265,50 @@ env.setRestartStrategy(RestartStrategies.noRestart())
...
+此处 Region 指以 Pipelined 形式进行数据交换的 Task 集合。也就是说，Batch 形式的数据交换会构成 Region 的边界。
+- DataStream 和 流式 Table/SQL 作业的所有数据交换都是 Pipelined 形式的。
+- 批处理式 Table/SQL 作业的所有数据交换默认都是 Batch 形式的。
+- DataSet 作业中的数据交换形式会根据 [ExecutionConfig]({{ site.baseurl }}/zh/dev/execution_configuration.html) 
+  中配置的 [ExecutionMode]({{ site.javadocs_baseurl }}/api/java/org/apache/flink/api/common/ExecutionMode.html)
+  决定。
+
+需要重启的 Region 的判断逻辑如下：
+1. 出错 Task 所在 Region 需要重启。
+2. 如果要重启的 Region 需要消费的结果数据 Partition 无法访问（丢失或损坏），产出该数据 Partition 的 Region 也需要重启。
```
Comment (gaoyunhaii) [inline docs/dev/task_failure_recovery.zh.md]: Do we need to translate the `ResultPartition` or we just leave it as in English ? I think it should be a terminology. Or we remove the word of `partition`.

File: docs/dev/task_failure_recovery.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (GJL) [reply docs/dev/task_failure_recovery.md]: ok

File: docs/dev/task_failure_recovery.md (hunk 20 above)
Comment (zhuzhurk) [reply docs/dev/task_failure_recovery.md]: Ok.

File: docs/dev/task_failure_recovery.md (hunk 21 above)
Comment (gaoyunhaii) [reply docs/dev/task_failure_recovery.md]: Got it, very sorry for the misunderstanding before.

File: docs/dev/task_failure_recovery.md (hunk 20 above)
Comment (gaoyunhaii) [reply docs/dev/task_failure_recovery.md]: Got it and agree with keeping it now.

File: docs/dev/task_failure_recovery.md (hunk 9 above)
Comment (GJL) [reply docs/dev/task_failure_recovery.md]: *"should"* can mean that a state is desirable or probable (see Section 1.1 or 2 in [1]).
However, restarting the region containing the failed task is a certain event. We can use will-future here, WDYT?

[1] https://www.lexico.com/en/definition/should

Comment (zhuzhurk) [reply docs/dev/task_failure_recovery.md]: I understand the concern now.
Agreed that we can use will-future.

File: docs/dev/task_failure_recovery.zh.md (hunk 22 above)
Comment (zhuzhurk) [reply docs/dev/task_failure_recovery.zh.md]: Agreed that we can refine this description.

File: flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java [hunk 23]
```diff
@@ -121,11 +125,10 @@
...
 					text("'full': Restarts all tasks to recover the job."),
 					text("'region': Restarts all tasks that could be affected by the task failure. " +
-						"The affected vertices consist of vertices in affected regions " +
-						"(pipelined connected sub-graphs), including " +
-						"the region of the failed task, " +
-						"downstream regions of affected regions, and" +
-						"producer regions of required but unavailable result partitions for this recovery.")
+						"More details can be found %s.",
+						link(
+							"https://ci.apache.org/projects/flink/flink-docs-master/dev/task_failure_recovery.html#restart-pipelined-region-failover-strategy",
```
Comment (GJL) [inline flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java]: At the moment there is no possibility to reference the site root from here. I would avoid linking the *master* docs here. The best alternative I can think of is to use a relative path: `../dev/task_failure_recovery.html#restart-pipelined-region-failover-strategy`. However, this assumes a specific directory structure.

Comment (zhuzhurk) [reply flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java]: Ok. We can use the relative path. No other good idea yet.

File: docs/dev/task_failure_recovery.md [hunk 24]
```diff
@@ -264,4 +267,54 @@ The cluster defined restart strategy is used.
...
+- The data exchange types in a DataSet job are determined by the 
+  [ExecutionMode]({{ site.javadocs_baseurl }}/api/java/org/apache/flink/api/common/ExecutionMode.html) 
+  which can be set through [ExecutionConfig]({{ site.baseurl }}/dev/execution_configuration.html).
+
+The regions to restart are decided as below:
+1. The region containing the failed task will be restarted.
+2. If a result partition is not available while it is required by a region that will be restarted,
+   the region producing the result partition will be restarted as well.
+3. If a region is to be restarted, all of its consumer regions will also be restarted. This is to guarantee
+   data consistency because nondeterministic processing or partitioning can result in that a partition
```
Comment (tillrohrmann) [inline docs/dev/task_failure_recovery.md]: nit: maybe replace `in that a partition ...` with `in different result partitions`.

Comment (zhuzhurk) [reply docs/dev/task_failure_recovery.md]: Ok. Agreed that the previous statement is cumbersome and the change is made.
But not sure whether the new statement may cause ambiguity to users, that it means one partition is different from its previous attempt, or one partition is different from another.

Comment (zhuzhurk) [reply docs/dev/task_failure_recovery.md]: Sorry, I thought I have published the previous reply but it is not.

**Comment by tillrohrmann** [comment]
Thanks a lot for addressing the comments @zhuzhurk. Merging this PR.
